### ISSUE TITLE
proposal: Remove buildx build resource limits arguments

### ISSUE THREAD

**Issue by Aposhian** [docker/buildx#644]
### Problem

`buildx build` accepts resource limit arguments for `--memory`, `--cpu-shares`, etc., but these are not currently being used in the code as noted in #292 #483. Furthermore, it may not make _sense_ to have these options for the generic `buildx build` command, since buildx can use builders that work in parallel and across multiple machines, which muddies the semantics of a simple option like `--memory`. This proposal is not intended to outline the future of resource limits in buildx/buildkit, only to clarify the current arguments and documentation to better match the current capabilities and behavior of buildx.

### Proposal

- [ ] Add a warning message whenever resource limit arguments are used on `buildx build` so that it is clear that they are being ignored. This preserves backward compatibility for people who use `buildx install`, but also does not mislead.
- [ ] Remove resource limit arguments from the documentation

**Comment by felipecrs** [comment]
I agree that the current approach is VERY misleading.

**Comment by ayk33** [comment]
Can we get some feedback on this issue?

**Comment by tonistiigi** [comment]
These flags do not appear in `build --help`. The only way to hit them is with old build scripts and as they do not change the outcome of the build there is no point to provide error for them.

Showing a warning when flag is set to custom value should be ok, but when we do decide to support the flag it would show an incorrect warning although the flag worked. Since this issue was created, we have already brought back a bunch of these flags.

**Comment by felipecrs** [comment]
@tonistiigi what do you mean they do not change the outcome? They for sure do depending on whatever happens during docker build:

https://github.com/felipecrs/oracle-docker-images/blob/main/OracleDatabase/SingleInstance/extensions/prebuiltdb/README.md#:~:text=If%20you%20know,memory%20option.

**Comment by tonistiigi** [comment]
> do not change the outcome

I meant the build result.

If you run commands that only work with certain configuration passed into them, then you shouldn't rely on the user passing in the correct value. Your build step could detect if there are not enough resources or use `ulimit` to limit itself. From the link you specified, doesn't that case actually use the 4G as a configuration variable that could be passed in with a `--build-arg` instead?

Regarding this specific case, setting memory limits for a single container is not a problem for buildkit. But this is not how most users would understand that flag. Instead they would think that this flag means limits for the whole build. In that case instead of passing a value with build request it makes more sense to add limits to the buildkit container or use the `--cgroup-parent` with limits that apply to all containers as a whole.

**Comment by felipecrs** [comment]
> From the link you specified, doesn't that case actually use the 4G as a configuration variable that could be passed in with a --build-arg instead?

No, because it's automatically calculated based on the amount of memory allocated for the container.

**Comment by tonistiigi** [comment]
> No, because it's automatically calculated based on the amount of memory allocated for the container.

That's what I meant. Instead of setting the value directly it expects the value to be set as a container limit and then inside the container, it goes in and reads what the limit was set to("automatic calculation").

**Comment by darklinght** [comment]
When using docker buildx build to create an image, it can cause the computer's CPU to max out. Are there any ways to limit CPU usage during the build process?

**Comment by Aposhian** [comment]
@darklinght There are a couple of strategies I have encountered, which come from outside of buildx itself. If you are using a Docker container builder, then you could set Docker resource limits on that container, or you can use cgroups to set resource limits.

**Comment by crazy-max** [comment]
Resource limits can now be set for CPU and memory using the `--resource` flag in build and the `resource` key in bake commands. This feature requires BuildKit v0.31.0+ and Dockerfile v1.25.0+.

More info: https://docs.docker.com/reference/cli/docker/buildx/build/#resource
